ably/realtime_presence_test.go

```diff
@@ -96,3 +104,38 @@ func TestRealtimePresence_Sync250(t *testing.T) {
 		t.Fatalf("members3: %v", err)
 	}
 }
+
+func TestRealtimePresence_EnsureChannelIsAttached(t *testing.T) {
+	beforePresenceGet := ^ably.StateEnum(0)
+	presTransitions := []ably.StateEnum{
+		ably.StateConnConnecting,
+		ably.StateConnConnected,
+		beforePresenceGet,
+		ably.StateChanAttaching,
+		ably.StateChanAttached,
+	}
+	rec := ably.NewStateRecorder(5)
+	opts := &ably.ClientOptions{
+		Listener:  rec.Channel(),
+		NoConnect: true,
+	}
+	app, client := testutil.ProvisionRealtime(nil, opts)
+	defer safeclose(t, client, app)
+	channel := client.Channels.Get("persisted:presence_fixtures")
+	if err := ably.Wait(client.Connection.Connect()); err != nil {
+		t.Fatal(err)
+	}
+	rec.Add(beforePresenceGet)
```

Comment:
> It's more subtle than that. I can't verify if rec.States() contains only [Connecting, Connected] since there may be states values buffered in the chan but not yet added to the slice. States() then will return [Connecting, Connected] and we'll be happy, but as soon as it returns the buffered in the channel states will then be added to the slice before presence.Get fires. What makes mark work is that it's sent via the same channel, that is it's enqueued at the end. Obviously it's not 100% accurate test, to be 100% sure we would need to introduce long timeouts before acknowledging the states, but in my opinion it's good enough for our case.

Since we are doing parallel processing here, it sounds sensible to have an expectation that `eventually` resolves the states should equal [Connecting, Connected]. In ginkgo that is very easy to do http://onsi.github.io/gomega/#eventually.